pkg/objstore/cos/cos.go

```diff
@@ -0,0 +1,423 @@
+func (b *Bucket) Delete(ctx context.Context, name string) error {
+	b.opsTotal.WithLabelValues(opObjectDelete).Inc()
+	if _, err := b.client.Object.Delete(ctx, name); err != nil {
+		return errors.Wrap(err, "delete cos object")
+	}
+	return nil
+}
+
+// Iter calls f for each entry in the given directory (not recursive.). The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	b.opsTotal.WithLabelValues(opObjectsList).Inc()
+
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, dirDelim) + dirDelim
+	}
+
+	for object := range b.ListObjects(ctx, dir, false) {
+		// this sometimes happens with empty buckets
+		if object.Key == "" {
+			continue
+		}
+		if err := f(object.Key); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (b *Bucket) getRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	b.opsTotal.WithLabelValues(opObjectGet).Inc()
+	if len(name) == 0 {
+		return nil, errors.Errorf("given object name should not empty")
+	}
+
+	opts := &cos.ObjectGetOptions{}
+	if length != -1 {
+		if err := setRange(opts, off, off+length-1); err != nil {
+			return nil, err
+		}
+	}
+
+	resp, err := b.client.Object.Get(ctx, name, opts)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := resp.Body.Read(nil); err != nil {
+		runutil.CloseWithLogOnErr(b.logger, resp.Body, "cos get range obj close")
+		return nil, err
+	}
+
+	return resp.Body, nil
+}
+
+// Get returns a reader for the given object name.
+func (b *Bucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	return b.getRange(ctx, name, 0, -1)
+}
+
+// GetRange returns a new range reader for the given object name and range.
+func (b *Bucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	return b.getRange(ctx, name, off, length)
+}
+
+// Exists checks if the given object exists in the bucket.
+func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
+	b.opsTotal.WithLabelValues(opObjectHead).Inc()
+	if _, err := b.client.Object.Head(ctx, name, nil); err != nil {
+		if b.IsObjNotFoundErr(err) {
+			return false, nil
+		}
+		return false, errors.Wrap(err, "head cos object")
+	}
+
+	return true, nil
+}
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	tmpErr := toErrorResponse(err)
+	if tmpErr.Code == "NoSuchKey" {
+		return true
+	}
+	if tmpErr.Response != nil && tmpErr.Response.StatusCode == http.StatusNotFound {
+		return true
+	}
+	return false
+}
+
+func toErrorResponse(err error) *cos.ErrorResponse {
+	switch err := err.(type) {
+	case *cos.ErrorResponse:
+		return err
+	default:
+		return &cos.ErrorResponse{}
+	}
+}
+
+func (b *Bucket) Close() error { return nil }
+
+type ObjectInfo struct {
+	cos.Object
```

Comment:
We can limit this to just key string right?

Reply:
em... agree.